File: packages/sdk-php/src/Resources/Events.php
<?php

declare(strict_types=1);

namespace ApexMail\Resources;

use ApexMail\Client;

/**
 * Query the email event log (delivery, opens, clicks, bounces, …).
 */
class Events
{
    public function __construct(private readonly Client $client) {}

    /**
     * List events with optional filters.
     *
     * Each event in the response may include an `envelope` key with:
     *   - from:      (string|null) Envelope MAIL FROM address
     *   - to:        (string[])    Envelope RCPT TO addresses
     *   - dkim:      (string)      Authentication result: pass|fail|neutral|none
     *   - spf:       (string)      Authentication result: pass|fail|neutral|none
     *   - dmarc:     (string)      Authentication result: pass|fail|neutral|none
     *   - timestamp: (string)      ISO 8601 delivery event timestamp
     *
     * @param array $options {
     *   @type string   $type       Event type, e.g. "message.delivered"
     *   @type string   $message_id Filter by specific message (email) ID
     *   @type string   $domain_id  Filter by sending domain
     *   @type string   $start      ISO 8601 start date
     *   @type string   $end        ISO 8601 end date
     *   @type int      $limit      Max results per page (default 50)
     *   @type int      $offset     Pagination offset
     *   @type string   $cursor     Cursor for cursor-based pagination
     * }
     */
    public function list(array $options = []): array
    {
        $query = http_build_query(array_filter([
            'type'      => $options['type']                                ?? null,
            'messageId' => $options['message_id'] ?? $options['messageId'] ?? null,
            'domainId'  => $options['domain_id']  ?? $options['domainId']  ?? null,
            'start'     => $options['start']                               ?? null,
            'end'       => $options['end']                                 ?? null,
            'limit'     => $options['limit']                               ?? 50,
            'offset'    => $options['offset']                              ?? 0,
            'cursor'    => $options['cursor']                              ?? null,
        ], static fn ($v) => $v !== null && $v !== ''));

        return $this->client->request('GET', '/v1/events' . ($query ? '?' . $query : ''));
    }

    /**
     * Retrieve all events for a specific sent message.
     *
     * @param string $messageId  The email ID returned by emails.send()
     */
    /**
     * Events for one message. The limit is exposed (default 100, server
     * cap) and the client's getNextCursor()/getLastResponseMeta() carry the
     * continuation when a retry storm produced more events than one page.
     */
    public function getByMessage(string $messageId, int $limit = 100): array
    {
        $limit = max(1, min(100, $limit));
        return $this->client->request(
            'GET',
            '/v1/events?messageId=' . urlencode($messageId) . '&limit=' . $limit
        );
    }

    /**
     * Get a single event by its ID.
     */
    public function get(string $eventId): array
    {
        return $this->client->request('GET', '/v1/events/' . urlencode($eventId));
    }

    /** Return aggregate event counts with optional filters. */
    public function stats(array $options = []): array
    {
        $query = http_build_query(array_filter([
            'type'      => $options['type']                                ?? null,
            'messageId' => $options['message_id'] ?? $options['messageId'] ?? null,
            'domainId'  => $options['domain_id']  ?? $options['domainId']  ?? null,
            'start'     => $options['start']                               ?? null,
            'end'       => $options['end']                                 ?? null,
        ], static fn ($v) => $v !== null && $v !== ''));

        return $this->client->request('GET', '/v1/events/stats' . ($query ? '?' . $query : ''));
    }

    /** Return event counts over time with optional filters. */
    public function timeseries(array $options = []): array
    {
        $query = http_build_query(array_filter([
            'type'      => $options['type']                                ?? null,
            'messageId' => $options['message_id'] ?? $options['messageId'] ?? null,
            'domainId'  => $options['domain_id']  ?? $options['domainId']  ?? null,
            'start'     => $options['start']                               ?? null,
            'end'       => $options['end']                                 ?? null,
            'interval'  => $options['interval']                            ?? null,
        ], static fn ($v) => $v !== null && $v !== ''));

        return $this->client->request('GET', '/v1/events/timeseries' . ($query ? '?' . $query : ''));
    }
}

File: packages/sdk-java/src/main/java/ee/apexmail/Events.java
package ee.apexmail;

import com.fasterxml.jackson.core.type.TypeReference;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * Query the email event log (delivery, opens, clicks, bounces, complaints, …).
 *
 * <p>Access via {@code client.events()}
 */
public final class Events {

    private final ApexMailClient client;

    public Events(ApexMailClient client) {
        this.client = client;
    }

    /**
     * List events with optional filters.
     *
     * <p>Supported filter keys: {@code type}, {@code messageId}, {@code domainId},
     * {@code start} (ISO 8601), {@code end} (ISO 8601), {@code limit}, {@code offset}.
     */
    public EventsListResponse list(Map<String, Object> options) {
        String q = options != null && !options.isEmpty()
            ? "?" + buildQuery(options) : "";
        return client.request("GET", "/v1/events" + q, null, EventsListResponse.class);
    }

    public EventsListResponse list() {
        return list(null);
    }

    /**
     * Get all events for a specific sent message (max 100 results).
     *
     * @param messageId  The email ID returned by {@code emails().send()}
     */
    public EventsListResponse getByMessage(String messageId) {
        return client.request("GET",
            "/v1/events?messageId=" + encode(messageId) + "&limit=100", null,
            EventsListResponse.class);
    }

    /**
     * Get a single event by ID.
     */
    public EventResponse get(String eventId) {
        return client.request("GET", "/v1/events/" + encode(eventId), null, EventResponse.class);
    }

    /** Return aggregate event counts with optional filters. */
    public Map<String, Object> stats(Map<String, Object> options) {
        String q = options != null && !options.isEmpty() ? "?" + buildQuery(options) : "";
        return client.request("GET", "/v1/events/stats" + q, null, new TypeReference<Map<String, Object>>() {});
    }

    public Map<String, Object> stats() {
        return stats(null);
    }

    /** Return event counts over time with optional filters. */
    public Map<String, Object> timeseries(Map<String, Object> options) {
        String q = options != null && !options.isEmpty() ? "?" + buildQuery(options) : "";
        return client.request("GET", "/v1/events/timeseries" + q, null, new TypeReference<Map<String, Object>>() {});
    }

    public Map<String, Object> timeseries() {
        return timeseries(null);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    private static String buildQuery(Map<String, Object> params) {
        StringBuilder sb = new StringBuilder();
        params.forEach((k, v) -> {
            if (v != null) {
                if (sb.length() > 0) sb.append('&');
                sb.append(encode(k)).append('=').append(encode(String.valueOf(v)));
            }
        });
        return sb.toString();
    }

    public record Event(
        String id,
        String messageId,
        String eventType,
        String recipientEmail,
        String timestamp
    ) {}

    public record EventsListResponse(java.util.List<Event> events, Map<String, Object> pagination) {}

    public record EventResponse(Event event) {}
}
